src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,128 @@
+import { ShaderGraphNode } from './shader-graph-node.js';
+
+var id = 0;
+
+/**
+ * @private
+ * @class
+ * @name pc.ShaderGraphBuilder
+ * @classdesc A Shader Graph Builder class
+ * @param {pc.Application} app - Application with shader graph core nodes registered that will be used to build the shader graph
+ */
+var ShaderGraphBuilder = function (app) {
+    id++;
+
+    this.app = app;
+    this._nodes = this.app.shaderNodes._nodes;
+
+    this._graph = new ShaderGraphNode();
+    this._graph.name = 'graphRoot_' + id;
+};
+
+ShaderGraphBuilder.prototype.constructor = ShaderGraphBuilder;
+
+Object.assign(ShaderGraphBuilder.prototype, {
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addParam
+     * @description adds a parameter input to graph
+     * @param {string} type - type of parameter
+     * @param {string} name - name of parameter
+     * @param {any} value - optional value of parameter
+     * @returns {any} returns the created input port
+     */
+    addParam: function (type, name, value) {
+        var ioPort = this._graph.addInput(type, name, value);
+        return ioPort;
+    },
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addNode
+     * @description creates and adds a core node to the shader graph and connects up inputs
+     * @param {string} name - name of parameter
+     * @param {object} [arguments] - first argument is name string, then 0-N output node ports or params to connect to new core node inputs - number and port type of arguments must match.

```

Comment:
"{object} [arguments]" - I think this might create typescript definition allowing a single argument only. It might need some way to indicate list of arguments, not sure what the right way to do this is.

Reply:
I saw this format used in JSDocs in another system - I will check (google it) if it right.

Reply:
https://jsdoc.app/tags-param.html#multiple-types-and-repeatable-parameters - seems it should be {...*}

Reply:
done